### ISSUE TITLE
K8SSAND-1427 ⁃ CRD upgrader needs changes for cass-operator 1.8.0 (CRD v1beta1 -> v1)

### ISSUE THREAD

**Issue by burmanm** [k8ssandra/k8ssandra#1152]
## Bug Report

**Describe the bug**
When updating the cass-operator from 1.7.1 to 1.8.0, the user needs to run the following command:

```console
kubectl patch crd cassandradatacenters.cassandra.datastax.com -p '{"spec":{"preserveUnknownFields":false}}'
```

Before the update of CRD goes through. This is required to support the Kube 1.22 environment. This isn't done by the CRD upgrader for cass-operator CRDs however, so we need to add this feature. This is required if the user wants to upgrade to newer Kubernetes version (this bug will not occur in kube 1.21 for example).

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1427) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1427
┆priority: Medium

**Comment by bradfordcp** [comment]
@burmanm is this ticket still valid?

**Comment by bradfordcp** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/K8ssandra-Grooming-623b197f1fcd4800110540ef/issues/k8ssandra/k8ssandra/1152?planning-poker) with ZenHub @burmanm @jsanda @Miles-Garnsey

**Comment by jsanda** [comment]
The description provides a work around. I am fine with that versus investing the time and effort into the code changes. Even if the code changes are minimal effort, there is still the additional time to test and then release.

**Comment by jsanda** [comment]
@burmanm what exactly is the problem if this change is not made? Do you expect that upgrading k8ssandra will fail?

For testing I decided on the following plan:

* Install k8s 1.21
* Install k8ssandra 1.3.3
* Upgrade k8ssandra to 1.5.0
* Upgrade k8s to 1.22

I first went through these steps **without** your changes so that I could see what problems I would hit. The upgrades succeeded without error. 

When I installed on k8s 1.21 the CRD has a non structural schema violation:

```yaml
status:
  acceptedNames:
    kind: CassandraDatacenter
    listKind: CassandraDatacenterList
    plural: cassandradatacenters
    shortNames:
    - cassdc
    - cassdcs
    singular: cassandradatacenter
  conditions:
  - lastTransitionTime: "2022-04-22T22:08:25Z"
    message: no conflicts found
    reason: NoConflicts
    status: "True"
    type: NamesAccepted
  - lastTransitionTime: "2022-04-22T22:08:28Z"
    message: the initial names have been accepted
    reason: InitialNamesAccepted
    status: "True"
    type: Established
  - lastTransitionTime: "2022-04-22T22:08:31Z"
    message: 'spec.preserveUnknownFields: Invalid value: true: must be false'
    reason: Violations
    status: "True"
    type: NonStructuralSchema
  storedVersions:
  - v1beta1
```
That warning remains after upgrading to k8s 1.22. The Helm upgrade succeeded without error. 

According to https://kubernetes.io/blog/2019/06/20/crd-structural-schema/,

```yaml
type: object
x-kubernetes-preserve-unknown-fields: true
```
is still valid with the newer `apiextensions.k8s.io/v1` api that will result in the old schema-less behavior.

Here is the CRD status after the helm upgrade:

```yaml
status:
  acceptedNames:
    kind: CassandraDatacenter
    listKind: CassandraDatacenterList
    plural: cassandradatacenters
    shortNames:
    - cassdc
    - cassdcs
    singular: cassandradatacenter
  conditions:
  - lastTransitionTime: "2022-04-22T22:08:25Z"
    message: no conflicts found
    reason: NoConflicts
    status: "True"
    type: NamesAccepted
  - lastTransitionTime: "2022-04-22T22:08:28Z"
    message: the initial names have been accepted
    reason: InitialNamesAccepted
    status: "True"
    type: Established
  storedVersions:
  - v1beta1
```
It is interesting to note that the stored version is v1beta1.

**Comment by burmanm** [comment]
This is what happens when one tries:

```console
➜  datastax kubectl apply -f https://raw.githubusercontent.com/k8ssandra/cass-operator/v1.7.1/docs/user/cass-operator-manifests.yaml
namespace/cass-operator created
serviceaccount/cass-operator created
secret/cass-operator-webhook-config created
Warning: apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
customresourcedefinition.apiextensions.k8s.io/cassandradatacenters.cassandra.datastax.com created
clusterrole.rbac.authorization.k8s.io/cass-operator-cr created
clusterrole.rbac.authorization.k8s.io/cass-operator-webhook created
clusterrolebinding.rbac.authorization.k8s.io/cass-operator-crb created
clusterrolebinding.rbac.authorization.k8s.io/cass-operator-webhook created
role.rbac.authorization.k8s.io/cass-operator created
rolebinding.rbac.authorization.k8s.io/cass-operator created
service/cassandradatacenter-webhook-service created
deployment.apps/cass-operator created
Warning: admissionregistration.k8s.io/v1beta1 ValidatingWebhookConfiguration is deprecated in v1.16+, unavailable in v1.22+; use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration
validatingwebhookconfiguration.admissionregistration.k8s.io/cassandradatacenter-webhook-registration created
➜  datastax cd cass-operator 
➜  cass-operator git:(master) ✗ kubectl apply --server-side --force-conflicts -k config/crd
customresourcedefinition.apiextensions.k8s.io/cassandratasks.control.k8ssandra.io serverside-applied
The CustomResourceDefinition "cassandradatacenters.cassandra.datastax.com" is invalid: spec.preserveUnknownFields: Invalid value: true: must be false in order to use defaults in the schema
➜  cass-operator git:(master) ✗
```

kubectl rejects the update (in 1.21).

**Comment by burmanm** [comment]
With the patches:

```console
➜  cass-operator git:(master) ✗ kubectl patch crd cassandradatacenters.cassandra.datastax.com -p '{"spec":{"preserveUnknownFields":false}}'
customresourcedefinition.apiextensions.k8s.io/cassandradatacenters.cassandra.datastax.com patched
➜  cass-operator git:(master) ✗ kubectl apply --server-side --force-conflicts -k config/crd                                            
customresourcedefinition.apiextensions.k8s.io/cassandradatacenters.cassandra.datastax.com serverside-applied
customresourcedefinition.apiextensions.k8s.io/cassandratasks.control.k8ssandra.io serverside-applied
➜  cass-operator git:(master) ✗ 
```

Even with the stored version staying at v1beta1, did the fields themselves update? Status' Tracked task should appear there.

**Comment by jsanda** [comment]
I performed the following steps:

* Install k8s 1.21.10
* Install k8ssandra 1.3.1
* Upgrade k8ssandra to 1.4.1. Worked without error.
* Upgrade to k8ssandra 1.5.0.
* Verified presence of new `trackedTasks` properties in CassandraDatacenter CRD

@burmanm I did reproduce the same behavior when upgrading with kustomize. I am not sure why the difference, but your steps aren't applicable to k8ssandra 1.x. Not sure what to do here.
